staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go

```diff
@@ -257,7 +264,16 @@ func (s *specProxier) handleGroupVersion(w http.ResponseWriter, r *http.Request)
 
 		for key := range apiServiceInfo.discovery.Paths {
 			if targetGV == key {
+				// The metrics logic is duplicated from InstrumentHandlerFunc in
+				// k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
+				// We cannot use the function to wrap the handler here
+				// because a single OpenAPI handler can serve multiple
+				// group versions and should be recorded differently.
+				delegate := &metrics.ResponseWriterDelegator{ResponseWriter: w}
+				w = responsewriter.WrapForHTTP1Or2(delegate)
+
 				apiServiceInfo.handler.ServeHTTP(w, r)
```

Comment:
can we not wrap this with `InstrumentHandlerFunc`? This looks like a good spot:
https://github.com/Jefftree/kubernetes/blob/c237ddb226fe3b9eb3bc909129af2405ff587b6c/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go#L115

Reply:
No because the same apiservice can serve multiple group versions. Unless we want to coalesce them into a single metric, they will need to be separated.

Reply:
I think group and version etc labels on the metrics indicate KRM api paths; the openapi data is still openapi formatted even if it happens to be describing a particular k8s GVR, so it shouldn't go into the metrics as if it is a request for that particular api. Maybe @logicalhan has more or different thoughts.

Reply:
Yeah, not planning to touch the group/version parts of the metric. Everything will be under subresource: eg: subresource: "/openapi/v3/apis/apps/v1"

Reply:
Overloading subresource is also kinda janky though.

Reply:
I don't think the gv need to be in the metrics at all. Don't use subresource or resource fields either, since those are about other things; these are not subresources in the KRM sense used for normal resources.

Reply:
Yeah I agree with @lavalamp

Reply:
OpenAPI documents are of different sizes for different paths so the response rate will greatly differ. I think we should differentiate them somehow in the metrics?

Reply:
For what purpose? What will we do with such information?

I think it is 100% fine if openapi v3 requests are high-variance in the metrics.

Reply:
Why does the variable nature of the openapi documents matter? I mean, from what I can tell about what you are saying, you're saying that depending on the GV, the payload size is going vary, which will in turn affect the overall latencies. But even so, do we really need to know which GVs are the slow ones? Can't we just intuitively infer that the slow ones are the biggest ones and vice versa? What's the point of this?

Reply:
> For what purpose? What will we do with such information?
> 
> I think it is 100% fine if openapi v3 requests are high-variance in the metrics.

Yeah, I'm basically at the same conclusion.

Reply:
hmm ok that makes sense. Is it fine sticking with subresource to indicate that the request belongs to openapi? Is there a better parameter that could be used?

Reply:
Is there a "non resource path" or something? that is where it should go if so

Reply:
No, we don't support a non-resource path, I'd probably just use resource.

Reply:
Maybe Han is more familiar but I don't think so. livez also uses subresource.

Reply:
Oh I was going to say the same thing as @lavalamp. I'd do whatever livez was doing.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go#L184 for reference.